### PR TITLE
Make C# ServerCallContext implementation agnostic

### PR DESCRIPTION
**Pull request by jtattermusch**
Progress towards allowing https://github.com/grpc/grpc-dotnet/issues/1

To be able to move ServerCallContext to a separate Grpc.Core.Api package (see https://github.com/grpc/grpc-dotnet/issues/23#issuecomment-453964930), we need to get rid of Grpc.Core implementation-specific details and make ServerCallContext basically and "empty shell".

This PR makes the necessary refactoring (moving the grpc-native specific logic to a different class) and also opens door for making ServerCallContext constructor public (should that be needed).

### CODE REVIEW

**Comment by jtattermusch** [comment]
CC @JunTaoLuo @JamesNK

File: src/csharp/Grpc.Core/Internal/CallSafeHandle.cs [hunk 1]
```diff
@@ -18,6 +18,7 @@
 using System.Diagnostics;
 using System.Runtime.InteropServices;
 using System.Text;
+using System.Threading;
```
Comment (JunTaoLuo) [inline src/csharp/Grpc.Core/Internal/CallSafeHandle.cs]: Is this change needed?

File: src/csharp/Grpc.Core/ServerCallContext.cs [hunk 2]
```diff
@@ -29,45 +30,49 @@ namespace Grpc.Core
     /// </summary>
     public class ServerCallContext
     {
-        private readonly CallSafeHandle callHandle;
+        private readonly object extraData;
```
Comment (JunTaoLuo) [inline src/csharp/Grpc.Core/ServerCallContext.cs]: Though this probably works but I'm not sure if it's the best way to separate the native types.

I assume this implementation is mostly concerned with back compat but ideally everything (types and methods) exposed publicly on `ServerCallContext` could have been defined on a `IServerCallContext` interface and a `NativeServerCallContext` would implement the interface with its own mechanism of storing the extra data and getters/setters for `WriteOptions`, `Peer`, `AuthContext`, etc.

Since we probably can't change all the generated code now to start accepting a `IServerCallContext` interface in the call handlers since that would be a major breaking change. I'm wondering if it's possible to make ServerCallContext an abstract class and move all the implementation into a `NativeServerCallContext` that extends from it?

I haven't tried this yet so I'm not sure about the ramifications but I'd like to explore that possibility before resorting to this mechanism with `objects` types and `Func` that you pass to the constructor to specify how to retrieve properties.

Though changing `ServerCallContext` from a POCO to an abstract base class is technically breaking, I think it's small enough of a breaking change to be worth it.

What do you think @shirhatti @jamesNK?

File: src/csharp/Grpc.Core/Internal/ServerCallContextExtraData.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            return ((ServerCallContextExtraData)extraData).serverResponseStream.WriteResponseHeadersAsync(headers);
+        };
+
+        readonly static Func<ServerCallContext, object, WriteOptions> ServerCallContext_WriteOptionsGetter = (ctx, extraData) =>
+        {
+
+            return ((ServerCallContextExtraData)extraData).serverResponseStream.WriteOptions;
+        };
+
+        readonly static Action<ServerCallContext, object, WriteOptions> ServerCallContext_WriteOptionsSetter = (ctx, extraData, options) =>
```
[the remark this answers is not in the document]
Comment (apolcyn) [reply src/csharp/Grpc.Core/Internal/ServerCallContextExtraData.cs]: to add, I could imagine setting `ServerCallContext_WriteHeadersFunc` is an instance property within the constructor

[the remark this answers is not in the document]
Comment (apolcyn) [reply src/csharp/Grpc.Core/Internal/ServerCallContextExtraData.cs]: other option, might be less cumbersome paramer-wise if `ServerCallContext` had e.g. a `GetExtraData` method on it

[the remark this answers is not in the document]
Comment (jtattermusch) [reply src/csharp/Grpc.Core/Internal/ServerCallContextExtraData.cs]: I considered both of these and decided not to do what you're suggesting:
- The extra "ServerCallContext" and "object" parameters in the lambdas is there to avoid unnecessary lambda allocations every time we create a ServerCallContext (the lambdas like ServerCallContext_WriteHeadersFunc need to be static.
- ServerCallContext.GetExtraData would have needed to be public and it would be visible to all the users writing the server side handlers - and I don't know that. If it was internal, every alternative grpc implementation using this API would need an [InternalsVisibleTo] annotation added to Grpc.Core.Api

File: src/csharp/Grpc.Core/Internal/CallSafeHandle.cs (hunk 1 above)
Comment (jtattermusch) [reply src/csharp/Grpc.Core/Internal/CallSafeHandle.cs]: removed.

File: src/csharp/Grpc.Core/ServerCallContext.cs (hunk 2 above)
Comment (jtattermusch) [reply src/csharp/Grpc.Core/ServerCallContext.cs]: You're right, this is all about backward compatibility, otherwise this change would be trivial.

Ad `IServerCallContext` interface:
- we definitely can't change `ServerCallContext` to `IServerCallContext` in the generated code. `ServerCallContext` is exposed in every server-side handler so a change would break a large portion of current users.
- using an interface for server call context is not right in general as there might be more members added to `ServerCallContext` over time (e.g. long time ago we added AuthContext) and adding new members to an interface is a breaking change (using ServerCallContext as a base class would be better as it could have default implementations for newly added members).

Ad changing `ServerCallContext` to an abstract class base:
- this is something we *could* consider, because currently there are no public constructors so the only instances of ServerCallContext in existence are the ones created in Grpc.Core or by Grpc.Core.Testing. I'd be worried about breaking the binary compatibility though (this has the potential to break all of the many existing nuget packages that depend on Grpc.Core).
See "adding virtual to a member" in https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#members
- I agree that just changing `ServerCallContext` to a base class is much simpler, but the solution I'm proposing is safer from the perspective of backward compatibility.

One possible solution would be to change ServerCallContext into a base class, leave the existing non-virtual members as is (adding "virtual" is breaking) and make them call a newly added protected virtual or abstract member (even abstract is fine as there are currently no accessible constructors outside of our control). These members can be then overridden by subclasses like 'NativeServerCallContext' that would provide the right behavior for each grpc implementation.
@apolcyn  @JunTaoLuo  WDYT?

**Comment by jtattermusch** [comment]
Based on the discussion, I updated the PR to what I believe is a superior solution:

- `ServerCallContext` has been turned into a abstract base class that has no fields and delegates all its functionality to `protected abstract` members (the original members needed to stay non-virtual to avoid breaking binary compatibility).
- `DefaultServerCallContext` implements all the original functionality of `ServerCallContext`
- Overall this leads to a more readable and flexible solution.
- The ServerCallContext now has a parameterless constructor that's `protected` and so instances of ServerCallContext can be used in alternative grpc implementations without requiring [InternalsVisibleTo] attribute.

@apolcyn  @JunTaoLuo  PTAL.

**Comment by JamesNK** [comment]
The latest changes work for us. Grpc.AspNetCore will need its own implementation of `ServerCallContext`. I suspect our implementation will be a wrapper over HttpContext https://github.com/grpc/grpc-dotnet/issues/35

As long as we can inherit from `ServerCallContext` and create new versions ourselves then we should be good.

**Comment by chwarr** [comment]
Would you be open to adding an `object UserData { get; set; }` property/field/method to `ServerCallContext`? I envision using such a field to [thread state between collaborating interceptors and calls](https://github.com/grpc/grpc/issues/15757#issuecomment-445053226) cheaper and more nicely bracketed than `AsyncLocal<T>`. I don't see a way to use my own concrete `ServerCallContext` child without wrapping all of call dispatching.

**Comment by jtattermusch** [comment]
> btw have you considered "Core" instead of "Internal" in the name of protected members? That is the name we (ASP.NET team) tend to use when we have a public method that wraps a non-public version of a member.
> 
> https://github.com/aspnet/AspNetCore/blob/7d4ae6049f48bc556369e11c8d740956fd8d0458/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs#L148-L155

Done.

**Comment by chwarr** [comment]
> adding the UserData property is something we can discuss (in fact I've heard more requests in that direction, we just haven't figured out the best design for such API). Anyway it should be done separately, outside of this refactoring PR.

Makes sense. I opened [Issue #17759, "Add public object UserData to C# ServerCallContext"](https://github.com/grpc/grpc/issues/17759) to capture the `UserData` need. I didn't see an existing issue; there are similar ones.
